### PR TITLE
	(FACT-1356) Allow executable facts to output Yaml/Json 

### PR DESCRIPTION
**Pull request by branan**
Previously, executable facts could only return simple key/value pairs. This code now attempts to parse executable output as Yaml, and only falls back to the key/value parsing if we do not get a useful result. Because Yaml is a superset of Json, we do not have to separately try to parse Json output.

This also moves the helper code for converting a Yaml object into facts into a new shared location where both the flat-file yaml resolver and the executable resolver can share it.

### CODE REVIEW

**Comment by glennsarti** [comment]
Do we need to be concerned about UTF8 being parsed correctly into a structured fact?

**Comment by branan** [comment]
@glennsarti As long as the system locale/encoding is set correctly we will be well-behaved. If a program gives us output that doesn't match that we can't really be trusted... but that was also the case for key/value facts

**Comment by glennsarti** [comment]
Ahh ok. So no worse off.

**Comment by branan** [comment]
Quick force-push for the linter. I'm not sure what happened on appveyor :/

**Comment by branan** [comment]
The windows failure is newline conversion. I'm going to go ahead and open a leatherman ticket to add a "this program outputs text, please convert its newlines" flag to the execution helpers.

**Comment by branan** [comment]
I've just pushed up a commit to use the newline conversion code I'm working on for leatherman. This will cause travis to fail until we get leatherman updated and released with that feature

**Comment by branan** [comment]
https://github.com/puppetlabs/leatherman/pull/208 contains the leatherman PR. We'll need to cut a leatherman release once that's landed in order to get this in.

**Comment by Iristyle** [comment]
As I commented on the other PR, I don't understand why the parsing doesn't use the appropriate line endings for a given platform?  Where does the failure occur - in libYaml?

File: lib/src/facts/external/execution_resolver.cc [hunk 1]
```diff
@@ -29,38 +32,51 @@ namespace facter { namespace facts { namespace external {
...
+                    // interpretation in those cases.
+                    resolved = true;
+                }
+            }  catch (YAML::Exception& ex) {
+                LOG_DEBUG("Could not parse executable fact output as YAML ({1})", ex.msg);
+            }
+
+            if (!resolved) {
+                std::vector<string> lines;
+                boost::split(lines, result.output, boost::is_any_of("\n"));
```
Comment (branan) [inline lib/src/facts/external/execution_resolver.cc]: @Iristyle The windows issue is this split. I could definitely just deal with the `\r` here, but it feels cleaner to me to let the I/O layer deal with that sort of platform-specific stuff. Maybe that just comes from me being used to `stdio` doing it, though.

**Comment by glennsarti** [comment]
You'd probably want a consistent line ending for all plats.  What would happen on Windows if you gave it only LFs instead of CRLFs

**Comment by Magisus** [comment]
I've opened https://github.com/puppetlabs/facter/pull/1449 with everything in this PR, plus additional tests and the leatherman bump. Closing this in favor of that one.

**Pull request closed without merging**